rio/src/RIO/File.hs

```diff
@@ -0,0 +1,345 @@
+        throwIO e
+    -- we want to truncate() if this is an open in WriteMode, but only if the
+    -- target is a RegularFile. ftruncate() fails on special files like
+    -- /dev/null.
+    when (iomode == WriteMode && fd_type == RegularFile) $ Device.setSize fD 0
+    HandleFD.mkHandleFromFD fD fd_type fp iomode False Nothing `onException`
+      (liftIO $ Device.close fD)
+
+-- | Opens a file using the openat C low-level API. This approach allows us to
+-- get a file descriptor for the directory that contains the file, which we can
+-- use later on to fsync the directory with.
+--
+-- @since 0.1.6
+openFileDurable :: MonadIO m => FilePath -> IOMode -> m (Fd, Handle)
+openFileDurable absFp iomode =  do
+  let dir = takeDirectory absFp
+      fp = takeFileName absFp
+
+  dirFd <- openDir dir
+  fileHandle <- openFileFromDir dirFd fp iomode
+  return (dirFd, fileHandle)
+
+-- | This sub-routine does the following tasks:
+--
+-- * It calls fsync and then closes the given handler (mapping to a temporal/backup filepath)
+-- * It calls fsync and then closes the containing directory of the file
+--
+-- These steps guarante that the file changes are durable.
+--
+-- @since 0.1.6
+closeFileDurable :: MonadIO m => Fd -> Handle -> m ()
+closeFileDurable (Fd dirFd) h =
+  liftIO $
+  finally
+    (do fileFd <- handleToFd h
+        fsyncDescriptor "File" (FD.fdFD fileFd) `finally` hClose h
+        fsyncDescriptor "Directory" dirFd)
+    closeDirectory
+  where
+    fsyncDescriptor name fd =
+      void $
+      throwErrnoIfMinus1 ("closeFileDurable - fsync(" <> name <> ")") $
+      c_safe_fsync fd
+    closeDirectory =
+      void $
+      throwErrnoIfMinus1Retry "closeFileDurable - close(Directory)" $
+      c_close dirFd
+
+toTmpFilePath :: FilePath -> FilePath
+toTmpFilePath filePath =
+    dirPath </> tmpFilename
+  where
+    dirPath = takeDirectory filePath
+    filename = takeFileName filePath
+    tmpFilename = "." <> filename <> ".tmp"
+
+
+
+handleToFd :: Handle -> IO FD.FD
+#if MIN_VERSION_base(4,9,0)
+handleToFd h =
+  case h of
+    HandleFD.FileHandle _ mv -> do
+      HandleFD.Handle__{HandleFD.haDevice = dev} <- readMVar mv
+      case cast dev of
+        Just fd -> return fd
+        Nothing -> error "not a file handle"
+    HandleFD.DuplexHandle {} -> error "not a file handle"
+# else
+handleToFd = HandleFD.handleToFd
+#endif
+
+-- | This sub-routine does the following tasks:
+--
+-- * It calls fsync and then closes the given handler (mapping to a temporal/backup filepath)
+-- * It renames the file to the original path (using renameat)
+-- * It calls fsync and then closes the containing directory of the file
+--
+-- These steps guarante that the file are durable, and that the backup mechanism
+-- for catastrophic failure is discarded after no error is thrown.
+--
+-- @since 0.1.6
+closeFileDurableAtomic ::
+     MonadIO m => FilePath -> Fd -> Handle -> m ()
+closeFileDurableAtomic filePath (Fd dirFd) fileHandler =
```

Comment:
There's a lot of overlapping logic with the above. Since this code is so sensitive to proper exception handling, I think refactoring to share the code would be warranted.